RFC: CStr, the dereferenced complement to CString

mzabaluev opened the pull request:
Change the `Deref` target of `CString` to a token type `CStr`.
Add helper functions and macros to produce `CStr` references out of static string data.

[Rendered](https://github.com/mzabaluev/rust-rfcs/blob/c-str-deref/text/0000-c-str-deref.md)

mzabaluev commented:
Cc @alexcrichton

dgrunwald commented:
:+1: to the general idea: some kind of `CStr` is necessary to allow writing efficient safe bindings around C libraries.

I think this could be added as a motivating usecase:
```c
// C library:
const char* f1(); // return value has static lifetime
void f2(const char*);
```

```rust
/// Safe rust bindings:
fn f1() -> &'static CStr { ... }
fn f2(s: &CStr) { ... }

// Caller using only safe code:
f2(f1())
```

If `CString` was used instead of `CStr`, `f1` would have to copy the string into the rust heap.
If `[u8]` was used instead of `CStr`, `f2` would have to unnecessarily validate that the string is null terminated.

I also like the 'irrelevantly-sized type': [my approach using a DST CStr](https://github.com/dgrunwald/rust-cpython/blob/master/src/cstr.rs) would involve an unnecessary `strlen` call in the binding for `f1`.

---

However, this proposal seems to be missing some details around the type:

* `static_c_str_from_bytes` and `static_c_str_from_str`: I don't think there's any reason these need to be restricted to `'static` lifetime, is there?
You should also mention that these functions would check the input slice and panic if it's not 0-terminated or contains non-terminating zeroes (there should probably be an `unsafe` unchecked version of these functions, too).

* There should be a way to convert from `*const libc::c_char` to `&CStr` without going through `&[u8]`.
* `Borrow`/`ToOwned` implementations should be provided to connect `CStr` with `CString`.
* Should there be functions like `as_bytes`, `as_bytes_with_nul` as on `CString`? (these would have to call `strlen`)
* In general, try to support all operations that are valid on `CString` and also make sense for non-owned strings.

mzabaluev commented:
@dgrunwald, thanks for your comments.

> `static_c_str_from_bytes` and `static_c_str_from_str`: I don't think there's any reason these need to be restricted to 'static lifetime, is there?

These address a special case: they are non-copying ways to get `&CStr` out of static values, which have to be null-terminated (unlike dynamic slices, which must not have NUL anywhere including the tail position). These functions can skip the O(N) check for interior NULs, because in most cases their parameter will be a literal where any interior `'\0'` characters/bytes would be evident. I will add this explanation to the RFC.

> There should be a way to convert from `*const libc::c_char` to `&CStr` without going through `&[u8]`.

In this proposal, I view `CStr` only as a way to pass data from Rust to C. For Rust code, the only useful thing to do with a returned C string pointer is to parse it into a slice, which can already be done with `c_str_to_bytes`/`c_str_to_bytes_with_nul` (Besides, these functions need some
[lifetime love](https://github.com/rust-lang/rfcs/pull/556)). I do share your concern about being able to feed a non-owned C string obtained from FFI to another FFI-wrapping interface without unnecessary copying, but there is uncertainty as to whether this will be needed often.
Perhaps there should be a raw builder for `&CStr`, like in my crate:
```rust
unsafe fn c_str_from_raw_ptr<'a, T: ?Sized>(ptr: *const libc::c_char,
                                          life_anchor: &'a T)
                                          -> &'a CStr
```

> `BorrowFrom`/`ToOwned` implementations should be provided to connect `CStr` with `CString`.

Yes, and trivially. This, however, is not essential to the RFC and can be added later.

> Should there be functions like `as_bytes`, `as_bytes_with_nul` as on `CString`? (these would have to call `strlen`)

I'm hesitant about these, and I don't quite see the purpose of `as_bytes` on `CString` to begin with. My expectation is that `CString` is constructed at the use site rather than being passed around in interfaces, so its user should have access to the original slice. In any case, `as_bytes` is not a good method name for what it takes to get a slice from a C string pointer, because methods with this name have negligible cost elsewhere in the core libraries. In my crate, similar functions are named `parse_as_bytes` to remind about the scanning cost. These methods will be also invokable on `CString` via auto-deref, so their peculiarity should be apparent.

An alternative is to make `CStr` a DST newtype over `[libc::c_char]`. This will preclude any thin wrappers that do not call `strlen` up front, like the RAII helper `OwnedCString` in my crate.

Diggsey commented:
How do you prevent instances of CStr being created on the stack without making it unsized?

mzabaluev commented:
@Diggsey, you can't safely move a borrowed non-`Copy` value, and `CStr` is only ever obtainable borrowed.

Diggsey commented:
It still seems like there's potential for problems: for example, if DerefMut is ever implemented for CString (not unlikely, as it doesn't enforce UTF8 afaict) then you'll have to special-case "swap" to prevent it being used with CStr. You're allowing the programmer to bypass any "Sized" constraints on type parameters: for example, "sizeof" would give incorrect results when applied to CStr.

What's really needed is an unsized type without fat references: maybe there should be an "Unsized" marker class, or a way to negatively implement the Sized trait for a type, which doesn't affect its in-memory representation.

mzabaluev commented:
@Diggsey:

> if DerefMut is ever implemented for CString (not unlikely, as it doesn't enforce UTF8 afaict)

FWIW `CString` enforces the no-NULs-except-last invariant. To maintain it with mutability, `CString` and `CStr` would have to grow some interesting infrastructure. I don't think, however, that facilitating mutation of C strings in Rust will be worthwhile. You'd likely need size changes, and for that going through `Vec` or `String` is a better option, I think. There is actually a case for `into_vec` method on `CString`, which I have on its alternative `CStrBuf` in my `c_string`.

> You're allowing the programmer to bypass any "Sized" constraints on type parameters: for example, "sizeof" would give incorrect results when applied to CStr.

Are there any examples of safe code demonstrating how this could be unintentionally misused? My understanding is that, as long as you can't safely initialize, copy, or move a value, its size should not be important.

I also use this technique when given a public view into an object that has a private part allocated alongside:
```rust
#[repr(C)]
pub struct Object {
    raw: ffi::RawObject,  // we know it's not all there is
    no_copy: std::marker::NoCopy
}

impl Object {
    // an example with a static ref, just for the sake of brevity
    pub fn get_static() -> &'static Object {
        unsafe {
            let ptr = ffi::raw_object_get_static();
            std::mem::copy_lifetime("", &*(ptr as *const Object))
        }
    }
}
```

Being able to opt out of `Sized` without implying DST would be nice, indeed. @nikomatsakis ?

Kimundi commented:
Example with safe code:

```rust
let a: &'a mut CStr = foo.as_c_str();
let b: &'b mut CStr = bar.as_c_str();

mem::swap(a, b);
```
This swaps the actual `CStr` values, which will incidentally result in swapping of the first bytes of each C string, which will lead to memory unsafety if one C string was empty and had its terminating null byte at first position.

quantheory commented:
If we don't allow any method to create a mutable `CStr` (which I think is actually for the best), the mem::swap issue goes away. Additionally, I think some of the unsized weirdness could be solved by instead having a representation like:

```rust
#[repr(C)]
pub struct CStringRef<'a> {
    raw: core::nonzero::NonZero<*const libc::c_char>,
    marker: std::marker::ContravariantLifetime<'a>,
}
```

That is, instead of trying to use an `&'a CStr` that's a normal reference, the `CStringRef<'a>` itself becomes a sized "reference-like" type into un-owned immutable data. The only issue is that `CStringRef` looks different from the various slice types, e.g. `&str`, `&[u8]`, and `&Path`. Personally, I think that this lack of syntactic sugar is OK, but otherwise it does seem like `CStr` would have to be made unsized.

If `CStr` was also primitive, the above `CStringRef` could be stripped of its lifetime parameter and used in the `Repr` impl for `CStr`, which would be a close analogy to how slices and `str` work, but making a new primitive type is probably going too far.

Diggsey commented:
CStr is not the only reason to want unsized types with thin references: the PIMPL idiom in C++ requires a similar ability. You can get away with just not having a dereferenced type, but it makes for some weird semantics: CStr and CString are both reference types, and yet CString dereferences to a CStr? Either that or you have to duplicate CStr functionality on a CString.

mzabaluev commented:
> instead of trying to use an `&'a CStr` that's a normal reference, the `CStringRef<'a>` itself becomes a sized "reference-like" type into un-owned immutable data. The only issue is that `CStringRef` looks different from the various slice types, e.g. `&str`, `&[u8]`, and `&Path`.

That will also work poorly with `Deref`, where the dereferenced value has to be returned by reference.

> otherwise it does seem like CStr would have to be made unsized.

Aside from the `size_of` issue which by itself is not very important, can any problems occur due to `CStr` having a bogus size? The same would apply for public views into larger private objects, as described above. When the ABI stabilizes and dynamic crates become the norm, I expect such implementation-hiding techniques to become more common between Rust crates as well, not only in interfacing with C.

quantheory commented:
> `CStr` and `CString` are both reference types, and yet `CString` dereferences to a `CStr`?

I don't think it's that strange, any more than `CString` deref-ing to `&CStr`.

> That will also work poorly with `Deref`, where the dereferenced value has to be returned by reference.

However, I think that this pretty much kills the idea. I don't know why I didn't think of that problem, since I've had the exact same issue with the `Index` trait. Not only can `Deref` not return a type with the right lifetime, it can't even be redefined to do so without higher-kinded types.

mzabaluev commented:
Some ideation on the forum: http://discuss.rust-lang.org/t/unsized-types-distinct-from-dsts/1385

Ericson2314 commented:
Unsized types definitely the way to go for this. Though I think dynamic sized [lib::char] wrapper enforcing C invariant could be useful too.

Your deallocation ideas should be subsumed by an allocator API, whenever we get it.

alexcrichton commented:
I think that in general this seems like a good direction to go in, but I want to lay out some specific concerns about the wording and what direction this is going in. I'd also like to apologize for being a little slow about getting to you on this!

---

Overall this RFC references DST and type sizes a lot, but I'm fairly confused as to why. For example the RFC as worded seems to imply that one of the downsides of deref-ing to `[libc::c_char]` *because* it is a DST, but I'm not sure why being a DST is a downside versus the choice of *what* the deref-ed to type is.

In my mind the concerns about whether the target of a `Deref` implementation is a DST or not is fairly irrelevant compared to *what* the target is. Would it be possible to re-word the motivation with less mentions of being a DST? I think some specific motivations are:

* There is no type in Rust which represents a C string allocated elsewhere (at 0 overhead). This is what a `CStr` type would satisfy.
* There is no way to take a C string in a Rust function without imposing an allocation somewhere (as today the way to do so is `&CString`).

You may have a few to add yourself, but it sounds like the richness of static information we're not currently expressing is a big concern.

---

I do want to point out that by having the target of `Deref` for `CString` be `CStr` we're throwing away statically known information (the length). I've been weighing how much this matters one way or another, and it seems like an OK downside for now.

Another piece to point out about `CString` is that making the "use this `CString` as a byte container" less ergonomic (one must call a method to get the byte slice). I think, however, that this is ok. I think the original RFC for the current `CString` failed to re-reconcile how `CString` is used which is to serve the one purpose of passing data from Rust to C, not for interpreting C data.

Can you perhaps mention these in the drawbacks section?

---

Some specific API critiques of mine:

> ```rust
impl CStr {
    pub fn from_static_bytes(bytes: &'static [u8]) -> &'static CStr { ... }
    pub fn from_static_str(s: &'static str) -> &'static CStr { ... }
}
```

I think that we may not want to expose these APIs for now. The `&[u8]` and `&str` types are so ubiquitous in Rust that it's easy to forget that you can only pass in types to these functions which have a trailing `0`. The `from_static_str` function is also somewhat unusual in that it reinterprets the meaning of the original string by chopping off the last byte (which is a little odd).

We can always add these APIs in the future, but there may be better methods of exposing them. For now a trip through `CString` would have to be done to create a `CStr`.

> `c_str!`

Would it be possible to leave this macro as a future extension for now? Adding new macros to the prelude is a pretty big step to take and I'm not sure that this would necessarily pull its weight just yet.

I'm a bit on the fence about this though, I just like being conservative!

> ```rust
impl CStr {
    pub unsafe fn from_raw_ptr<'a>(ptr: *const libc::c_char) -> &'a CStr
    { ... }
}
```

I think this would be better named `from_raw` as the `ptr` aspect is already implied by the pointer type being taken.

> ```rust
impl CStr {
    pub fn parse_as_bytes(&self) -> &[u8] { ... }
}
```

I think this would be better named `to_bytes` as there's not necessarily a whole lot of parsing going on and the use of `to` clearly signals that work is being done (aka `strlen`).

---

Overall I think this is a pretty good direction for `CString` to go in. It cuts back the exposed API *even further* which to me is always a good sign.

dgrunwald commented:
@alexcrichton: Did you take a look at my [DST CStr implementation](https://github.com/dgrunwald/rust-cpython/blob/master/src/cstr.rs)? (this RFC is loosely based on my implementation)

On the details:
> from_static_bytes, from_static_str:
Agreed, we don't need to expose these yet.
If someone really needs to create a `CStr` without allocating, it's still possible to use `from_raw_ptr`.

I think these were much less problematic in my original implementation:
```rust
pub fn from_bytes_with_nul(v: &[u8]) -> &CStr { ... }
pub unsafe fn from_bytes_with_nul_unchecked(v: &[u8]) -> &CStr { ... }
```
I didn't have a `c_char` version, since `CString` doesn't have that either.

> c_str
The macro is really useful, but I agree it shouldn't be in the `std` prelude. The macro will have to wait until it's possible to selectively import it.
But we should make it easy for users to write their own local copy of the macro if they need it -- which would be an argument for having `from_bytes_with_nul`.

> I think this would be better named from_raw as the ptr aspect is already implied by the pointer type being taken.

Given that it's the opposite as `as_ptr`, shouldn't it be named `from_ptr`? Also, the unconstrained output lifetime seems dangerous. `std::slice::from_raw_buf` accepts a reference to the pointer to keep the output lifetime constrained; maybe we should do the same here?

mzabaluev commented:
@dgrunwald:
> Also, the unconstrained output lifetime seems dangerous. std::slice::from_raw_buf accepts a reference to the pointer to keep the output lifetime constrained; maybe we should do the same here?

This is discussed in #556. I have updated the text to refer to that proposal.

mzabaluev commented:
Thanks @alexcrichton, I have addressed most of your concerns.

```rust
    impl CStr {
        pub fn from_static_bytes(bytes: &'static [u8]) -> &'static CStr { ... }
        pub fn from_static_str(s: &'static str) -> &'static CStr { ... }
    }
```

> I think that we may not want to expose these APIs for now. The &[u8] and &str types are so ubiquitous in Rust that it's easy to forget that you can only pass in types to these functions which have a trailing 0. The from_static_str function is also somewhat unusual in that it reinterprets the meaning of the original string by chopping off the last byte (which is a little odd).

These address an often-used special case: using static data, primarily literals, as C strings. Chopping off the trailing byte is a practical concession to enable zero-overhead conversion; the `c_str!` macro would have been the most practical way for using these functions, hiding the termination kink. Because the lifetime is `'static`, these functions are safe besides the runtime check for trailing `\0`. The rest of the string needs not be checked, because an interior `\0` in this case is most likely caused by a programmer error, and this will not result in undefined behavior. `from_static_str` can be dropped because it's trivially implemented.

mzabaluev commented:
> Overall I think this is a pretty good direction for CString to go in. It cuts back the exposed API even further which to me is always a good sign.

I have added a question regarding `c_str_to_bytes`*. Now that we want to change the lifetime semantics, we might as well remove those functions in favor of composition of the APIs discussed here.

alexcrichton commented:
@dgrunwald 

> @alexcrichton: Did you take a look at my DST CStr implementation? (this RFC is loosely based on my implementation)

I had not! I now have though, thanks for the pointer. It seems to take a fairly significantly different direction than this RFC though in that `&CStr` is a fat pointer which means that the length must always be known (as opposed to this RFC which defers that computation to later).

> ```rust
pub fn from_bytes_with_nul(v: &[u8]) -> &CStr { ... }
pub unsafe fn from_bytes_with_nul_unchecked(v: &[u8]) -> &CStr { ... }
```

Hm these are definitely interesting functions to have. It may be worthwhile to have an `_unchecked` variant somewhere, but we'd still have to worry about silent truncation by taking an arbitrary `v: &[u8]` (we'd need to scan for internal 0s).

> Given that it's the opposite as as_ptr, shouldn't it be named from_ptr?

I'd be down with that!

---

@mzabaluev 

> Thanks @alexcrichton, I have addressed most of your concerns.

These changes look great to me, thanks!

> These address an often-used special case: using static data, primarily literals, as C strings.

My one concern about this is the difference in semantics of `CString` vs `CStr` if this were to pass. `CString::from_slice` will panic if it finds an interior nul byte, but `CStr::from_static_bytes` would accept it and then later possibly silently truncate the value.

If we are ok with the truncation semantics, then I think `CString` should be modified as well. If we are not ok with the truncation semantics, I think this would optimally be implemented as a syntax extension. The syntax extension would probably be called `c_str!` and could statically verify there are no internal nul bytes and would then append one at the end. The expansion would look like:

```rust
unsafe { ::std::ffi::CStr::from_ptr(concat!("foo", "\0").as_ptr() as *const _) }
```

I think in general the exact guarantees the signatures are providing would influence how comfortable I am with them.

> I have added a question regarding c_str_to_bytes*.

I think we'd definitely want to deprecate them in favor of methods on `CStr` itself.

aturon commented:
@mzabaluev 

I'm late to this party, but this is a great RFC! I love the idea of introducing a slice type for null-terminated data. I think the various motivations are well-covered in the RFC and this thread, and I buy them. This will really tighten up the `CString` API.

That said, I agree with the original stance that @mzabaluev took that `CStr` really "should" be unsized, and at some point I'd like to pursue the "truly unsized types" RFC or something like it to get there. There are a number of places in the libraries and language where unsized types get somewhat special treatment, and it'd be good to be able to plug into that. (I can elaborate if people are curious what I mean.)

So that said, I'm interested in *some* means of exposing this type as unsized from the get-go. (Whereas the main proposal here exposes that `CStr` is sized.) Personally, I think making this a newtype wrapper for `[libc::c_char]` is the way to go: it gives us the correct semantics at a *temporary* runtime cost. Eventually, if/when we add "truly" unsized types or some other support, we can transition the representation from fat to thin pointer *without breaking code*. Alternatively, if we decide not to land such support, we should be able to transition to the main proposal here, again without breakage.

mzabaluev commented:
@aturon, if `CStr` is temporarily made a DST, could there be any usages of `&CStr` that will break when that becomes a thin reference?
Should `CStr` be left unstable until the question of unsized types gets resolved?

aturon commented:
@mzabaluev 

> if `CStr` is temporarily made a DST, could there be any usages of `&CStr` that will break when that becomes a thin reference?

I think only for unsafe code that depended on the exact representation. I will try to get some second opinions on this though, it's an important question.

> Should `CStr` be left unstable until the question of unsized types gets resolved?

I don't think so -- we want to stabilize `CString` soon, and I would really like to use the overall approach you're suggesting.

alexcrichton commented:
I think in general using Rust types in FFI is pretty hazardous and I don't think either way that we should start defining FFI functions like:

```rust
extern {
    fn foo() -> &'static CStr;
}
```

I would be ok changing the runtime representation of `CStr` at a future date to become a thin pointer instead of a fat pointer and mandating that the only stable method to get the raw representation is `.as_ptr()`

mzabaluev commented:
> I think only for unsafe code that depended on the exact representation.

That's not an issue for me; I don't think any code that would e.g. transmute the reference into a slice could expect any guarantees on backwards compatibility.

What's the general stance on binary compatibility in Rust 1.0 timeframe? Does anyone care?

aturon commented:
@mzabaluev 

> What's the general stance on binary compatibility in Rust 1.0 timeframe? Does anyone care?

Completely off the table. ABI compatibility is going to be a very long road.

mzabaluev commented:
@alexcrichton I don't suggest `&CStr` to be used in the FFI signatures themselves. It's only meant for safe wrapper functions.

`.as_ptr()` could be provided by trait `AsRaw` from the unsized types RFC, if that part of the RFC would land in Rust 1.0 in anticipation of the eventual support for `!Sized`.

alexcrichton commented:
@mzabaluev would you be ok with the `[libc::c_char]` internal representation that @aturon mentioned? We'd keep the same API that's already described as it would allow us to move to a thin pointer in the future.

mzabaluev commented:
@alexcrichton Yes, I'm convinced. I will update the RFC soon.

mzabaluev commented:
I've got another opportunistic change to suggest. Having the conversion functions panic on inner NULs is a little... severe. How about `CString::from_slice` return `Result<CString, NulError>`, and `CString::from_vec` return a `Result` with a move-friendly error around `NulError`, similar to how the conversions from UTF-8 fail with `Utf8Error` and `IntoUtf8Error`? I have prototyped this in cargo crate `c_string`.

aturon commented:
@mzabaluev I would be in favor of such a change. We'll probably still panic in common cases when using `CString` within `std`, but this way other users of `CString` can make their own choice.

alexcrichton commented:
I personally see the panic on inner nul as a contract of the `CString` API. I vaguely remember in the past that I tried to move to a `Result` but it just got so painful so quickly that I abandoned the efforts. I would vote for continuing to panic instead of returning a `Result`, but it's also not the end of the world because I can always just write my own function :)

mzabaluev commented:
@alexcrichton A potential problem is, it is actually _very rare_ that a real-life string has NUL characters. So developers may be blissfully forgetful of the panic until it hits their software in the field. DoS possibilities galore.

alexcrichton commented:
I personally see the fact that it's so rare as a justification for panicking as it's otherwise that much more difficult to deal with for *everyone* using `CString`. I do agree though that almost no one will consider the fact that `&str` can have interior nul bytes (I certainly don't consciously think about it).

aturon commented:
I posted a [discuss thread](http://internals.rust-lang.org/t/panic-in-cstring-from-slice-and-friends/1580/2) on the topic of panicking here.

mzabaluev commented:
I have updated the proposal to @aturon's suggestions, adding a disclaimer that the inner slice is not guaranteed to have any particular size. This way we can still avoid the `strlen()` on construction and discourage transmute-happy code.

aturon commented:
@mzabaluev Thanks again for this RFC! After discussion on this thread, on the discuss forum, and amongst the team, it seems that almost everyone is in favor of this change. We have decided to move forward, and try to land these revisions ahead of alpha2 if possible.

[Tracking issue](https://github.com/rust-lang/rust/issues/22469)
